---
layout: default
search_title: "Using the SISTR Pipeline"
description: "A guide on using the SISTR Pipeline."
---

# Using the SISTR Pipeline

This guide describes how to make use of the [*Salmonella in-silico* Typing Resource][sistr-web] (SISTR) pipeline within IRIDA. This pipeline enables the identification of the serovar and cgMLST types for *Salmonella* whole genome sequencing (WGS) data through comparisons with a large database (10,000+) of *Salmonella* genomes within [NCBI][].

* TOC
{:toc}

## Pipeline Overview

The SISTR pipeline that is implemented within IRIDA makes use of the following steps to translate WGS data into typing information:

1. Paired-end reads are merged with [FLASH][].
2. Merged and un-merged reads are assembled *de novo* using [SPAdes][].
3. Low-coverage and small contigs are removed from the generated assembly.
4. The assembled genome is passed to [sistr_cmd][], a command-line program for comparing genomes against the SISTR database.

## Running the Pipeline

The SISTR pipeline can bet set up to run using two separate methods.

### 1. Automated Execution

The SISTR pipeline can be set to run automatically on upload of new sequencing data to particular projects. This can be set either on the creation of a new project, or from the project settings page.

a. **Creation of New Project**

   On creation of a new project, the below option can be selected to enable the automated execution of SISTR on upload of new data to this project.

   ![new-project-sistr.png][]

b. **Project Settings**

   Automated SISTR analysis can also be enabled (or disabled) after a project is created from the project **Settings** page.

   ![project-settings-sistr.png][]

If automated execution of SISTR has been enabled for a project, then a SISTR analysis will be scheduled for execution on the upload of new sequencing data.  The results are accessible from the particular sample page.

![sistr-typing-sample.png][]

Clicking the **Automated SISTR Typing** link brings you to the appropriate analysis page for SISTR.

![sistr-typing-status.png][]

### 2. Manual Execution

To execute SISTR manually, please refer to the [IRIDA/SISTR Tutorial][].

## SISTR Results

### Status of `PASS`

A successfull SISTR run (with status of `PASS`) should produce the following page as output.

![sistr-results.png][]

### Status of `WARNING`

A SISTR run with a status of `WARNING` should produce the below output.

![sistr-results-warning.png][]

### Status of `FAIL`

An unsuccessfull SISTR run (with status of `FAIL`) should produce the following as output.

![sistr-results-fail.png][]

### Report

Interpretation of the produced output is as follows:

#### 1. SISTR Information

Basic information on the sample and quality of the SISTR results.

* **Sample Name**: The name of the sample used within this analysis.
* **Quality Control Status**: A value of `PASS` or `FAIL` depending on the quality of the input genome as determined by SISTR.
* **Quality Control Details**: In the case of a status of `WARNING` or `FAIL`, the particular reason why these results did not pass.

#### 2. Serovar Predictions

The *in silico* serovar predictions generated from SISTR.

* **Serovar (overall)**: The overall serovar prediction from both the antigen and cgMLST methods employed by SISTR.
* **Serovar (antigen)**: The serovar prediction generated by looking for the presence of specific antigen genes.
* **Serovar (cgMLST)**: The serovar prediction generated by comparing the cgMLST type of this genome to the SISTR database.
* **Serogroup**: The serogroup, identified based on the detected alleles of the *wzx* and *wzy* genes.
* **H1**: The H1 antigen prediction based off of the detected *fliC* allele. If absent, a `-` is reported.
* **H2**: The H2 antigen based of off the detected *fljB* allele. If absent, a `-` is reported.
* **O-antigen**: The O-antigen prediction (inferred from the overall serovar prediction).

#### 3. cgMLST330

The results of additional predictions made using the SISTR cgMLST330 schema.

* **Subspecies**: The subspecies identified by comparing the genome to the cgMLST database in SISTR.
* **Matching genome name**: The name of the closest matching genome in the SISTR database.
* **Alleles matching genome**: The number of alleles that have a sufficient match (non-missing, non-partial/truncated) to the closest genome in the SISTR database.
* **Percent matching**: The percent of alleles perfectly matching the closest genome (e.g., _100% * Alleles matching genome/330_).
* **cgMLST Sequence Type**: A SISTR Sequence Type number associated with the particular cgMLST profile. A sequence type will only be assigned when the full complement of 330 cgMLST genes are found (whether a perfect match or an imperfect but non-truncated match).

#### 4. Mash

The results of predictions made through comparisons using the software [Mash][]. Generally, cgMLST results are preferred over Mash.

* **Subspecies**: The subspecies identified by comparing the genome to the SISTR database.
* **Serovar**: The serovar identified using Mash.
* **Matching genome name**: The name of the closest matching genome in the SISTR database identified using Mash.
* **Mash distance**: The Mash distance to the closest matching genome. This approximates the mutation rate.

### Output Files

In addition to the report, the SISTR pipeline produces the following files available for download.

* `sistr-cgmlst-profiles.csv`: A comma-separated values file listing the cgMLST allelic profile for the genome.
* `assembly-status-with-repeats.txt`: Some basic statistics on the *de novo* assembly used by SISTR.
* `sistr-alleles-out.json`: A JSON file containing details on each of the allele calls.
* `contigs-with-repeats.fasta`: The set of contigs generated from the *de novo* assembly and used by SISTR.  This file could be uploaded to the [SISTR web application][sistr-web] for additional details and visualization of the results.
* `sistr-predictions.json`: The SISTR prediction results used to generate the SISTR report.
* `sistr-novel-alleles.fasta`: A list of any novel alleles that were detected by SISTR.

More information on the interpretation of these files is available on the [sistr_cmd][] page.

[sistr-web]: https://lfz.corefacility.ca/sistr-app/
[sistr-results.png]: {{ site.baseurl }}/images/tutorials/common/pipelines/sistr-results.png
[NCBI]: https://www.ncbi.nlm.nih.gov/
[FLASH]: https://doi.org/10.1093/bioinformatics/btr507
[SPAdes]: http://bioinf.spbau.ru/spades
[sistr_cmd]: https://github.com/peterk87/sistr_cmd/
[Mash]: https://mash.readthedocs.io/en/latest/
[project-settings-sistr.png]: images/project-settings-sistr.png
[new-project-sistr.png]: images/new-project-sistr.png
[sistr-typing-sample.png]: images/sistr-typing-sample.png
[sistr-typing-status.png]: images/sistr-typing-status.png
[sistr-results-fail.png]: images/sistr-results-fail.png
[sistr-results-warning.png]: images/sistr-results-warning.png
[IRIDA/SISTR Tutorial]: ../../tutorials/sistr/
